### DISCUSSION TITLE
Welcome to home_assistant_tgtg Discussions!

### DISCUSSION THREAD

**Discussion by Chouffy** [Chouffy/home_assistant_tgtg#6] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by Thundervf** [comment]
Hello. How do you get the access token using hacs?

**Reply by tjorim** [reply]
Please read the documentation: https://github.com/Chouffy/home_assistant_tgtg?tab=readme-ov-file#1-get-access-tokens

"This is to be executed outside of Home Assistant, i.e. on your local machine."

**Reply by Thundervf** [reply]
I already know that. But if I'm asking it it's because I don't understand how to get it.

**Reply by tjorim** [reply]
This part has nothing to do with HACS: the only thing HACS does is downloading this custom integration to your Home Assistant installation. On your pc: install python, install the tgtg python library and run the tgtg_get_tokens script.

**Reply by samueltupman-ui** [reply]
Im having an issues with the https://geo.captcha-delivery.com/captcha error appearing once I enter my email address... is there a work around or something I need to change to make it bypass this? Thanks in advanced

**Comment by frankm000** [comment]
thank you very much for this project. it has helped me a lot. unfortunately, however, the wrong time was displayed which is probably related to the time zone. Here is the part i changed in the sensor.py: 

from __future__ import annotations
from datetime import datetime, timedelta


    def update(self) -> None:
        global tgtg_client
        self.tgtg_answer = tgtg_client.get_item(item_id=self.item_id)

        # time shift 2 hours forward for pickup_start
        pickup_start_time = self.tgtg_answer.get("pickup_interval", {}).get("start")
        if pickup_start_time:
            try:
                pickup_start_time = datetime.fromisoformat(pickup_start_time)
                pickup_start_time += timedelta(hours=2)
                pickup_start_time = pickup_start_time.isoformat()
                self.tgtg_answer["pickup_interval"]["start"] = pickup_start_time
            except ValueError:
                pass
        
        # time shift forward by 2 hours for pickup_end
        pickup_end_time = self.tgtg_answer.get("pickup_interval", {}).get("end")
        if pickup_end_time:
            try:
                pickup_end_time = datetime.fromisoformat(pickup_end_time)
                pickup_end_time += timedelta(hours=2)
                pickup_end_time = pickup_end_time.isoformat()
                self.tgtg_answer["pickup_interval"]["end"] = pickup_end_time
            except ValueError:
                pass

        self.store_name = self.tgtg_answer["display_name"]
        self.item_qty = self.tgtg_answer["items_available"]

**Reply by philipp** [reply]
Time zones are a difficult problem in general.
It looks like the timestamps that come back are always UTC (and end in "Z")
It's probably tolerable for parts of the year in parts of Europe.  But anywhere else in the world, the timestamps are pretty useless as-is.
Of course it's quite possible that most people just use it to see if *anything* is visible, and if so, go and get it.

Here's a minimalist approach to fixing the times:
Add this with all the other from/import statements at the top of sensor.py
`from dateutil.parser import parse`
And then wrap all three of the lines that transfer the timestamp data in
`parse(` and `).astimezone().isoformat()`

So the whole block (starting around line 180) looks like this
```
        if "pickup_interval" in self.tgtg_answer:
            if "start" in self.tgtg_answer["pickup_interval"]:
                data[ATTR_PICKUP_START] = parse(self.tgtg_answer["pickup_interval"]["start"]).astimezone().isoformat()
            if "end" in self.tgtg_answer["pickup_interval"]:
                data[ATTR_PICKUP_END] = parse(self.tgtg_answer["pickup_interval"]["end"]).astimezone().isoformat()
        if "sold_out_at" in self.tgtg_answer:
            data[ATTR_SOLDOUT_TIMESTAMP] = parse(self.tgtg_answer["sold_out_at"]).astimezone().isoformat()
```
This will convert the timestamps to the timezone of the server that is doing the processing.  That *should* be good enough for many people.
Since it doesn't make any other modifications, the same template that is already in use should just do the right thing.

Fancier changes can be left to the reader, or myself if I get ambitious.

**Reply by philipp** [reply]
Looking more closely, could presumably also use `datetime.fromisoformat()` instead of `parse()`
No idea which one to prefer.  But I like the `astimezone()` as a more official way of handling it (can always create a timezone to pass in that represents your particular offset if you don't want the server timezone)

**Reply by frankm000** [reply]
> Looking more closely, could presumably also use `datetime.fromisoformat()` instead of `parse()` No idea which one to prefer. But I like the `astimezone()` as a more official way of handling it (can always create a timezone to pass in that represents your particular offset if you don't want the server timezone)

WOW!, thanks alot for the help, should be a inherent part of the component! Thanks!

**Reply by tjorim** [reply]
Actually Home Assistant integrations are always supposed to feed data as UTC. I was looking into it as I also noticed it displayed the time with a 2 hours difference. However, this was only on my dashboard (with a custom template) while the attributes in the entity shows it correctly. Apparently the template (which is also documented in this repo) just cuts up the string with `pickup_end[11:16]`. Proper timestamp handling would use a a proper filter/function: https://www.home-assistant.io/docs/configuration/templating/.
